subprojects/dependency-management/src/main/java/org/gradle/api/internal/artifacts/ivyservice/resolveengine/graph/DependencyGraphBuilder.java

```diff
@@ -179,6 +187,40 @@ private void performSelection(final ResolveState resolveState, ComponentState mo
         }
     }
 
+    private boolean tryCompatibleSelection(ResolveState resolveState, ComponentState moduleRevision, ModuleIdentifier moduleId, String version, ModuleResolveState module) {
+        ComponentState selected = module.selected;
+        Set<SelectorState> selectedBy = moduleRevision.allResolvers;
+        if (selected != null && selected != moduleRevision) {
+            if (moduleRevision.allSelectorsAgreeWith(selected.getVersion())) {
+                // if this selector agrees with the already selected version, don't bother and pick it
+                return true;
+            }
+
+            // Now check if all selectors from the module agree with the fact we should select this version
+            boolean allAccept = true;
+            boolean atLeastOne = false;
+            for (SelectorState selector : module.selectors) {
+                if (selectedBy.contains(selector)) {
+                    continue;
+                }
+                atLeastOne = true;
+                VersionSelector previousVersionSelector = selector.versionSelector;
+                if (previousVersionSelector == null || !previousVersionSelector.canShortCircuitWhenVersionAlreadyPreselected() || !previousVersionSelector.accept(version)) {
+                    allAccept = false;
+                    break;
+                }
+            }
+            // all selectors agree, let's pick it
+            if (atLeastOne && allAccept) {
+                resolveState.deselectVersionAction.execute(moduleId);
+                module.softRestart(moduleRevision);
```

Comment:
Why do we need the restart here? Why is setting all other versions to `New` not enough?